feat: mrazadar/PON-82/subscription-api-mock-and-design-changes

mrazadar opened the pull request:
## Description
<!--
Describe what this pull request changes, and why these changes were made. How will these changes affect other people, installations of edx, etc.?

Please include links to any relevant ADRs, design artifacts, and decision documents. Make sure to document the rationale behind significant changes in the repo, per [OEP-19](https://open-edx-proposals.readthedocs.io/en/latest/oep-0019-bp-developer-documentation.html), and can be linked here.

Useful information to include:
- Which edX user roles will this change impact? Common user roles are "Learner", "Course Author", "Developer", and "Operator".
- Include screenshots for changes to the UI (ideally, both "before" and "after" screenshots, if applicable).
- Provide links to the description of corresponding configuration changes. Remember to correctly annotate these changes.
-->

This pull request includes many design and api data mock changes to new subscription checkout page. The list of the change set is given below: 

- Added new shimmers loading states as per design
- Added new resubscribe button state
- Added subscription/data layer to do subscription specific checks
- Added dynamic content rendering via subscription redux layer
- Optimized class based components to functional components for subscription feature
- Removed old unused code
- Implemented SubscriptionCheckout component
- Refactor some old code
- Improve responsive styles
- Implemented subscription confirmation
- Refactored better names for subscription feature


![Screen Shot 2023-03-15 at 5 56 55 PM](https://user-images.githubusercontent.com/5585262/225316099-e6ceb0c4-d082-43c4-b067-088a5d03f40d.png)
![Screen Shot 2023-03-15 at 6 08 14 PM](https://user-images.githubusercontent.com/5585262/225318124-ee1c3c84-9b90-45d4-aa5e-9cf62b297356.png)

mrazadar committed:
feat: PON-82 Subscription design changes-v2

feat: PON-82 Subscription Confirmation Modal

feat: PON-82 updated skeleton

feat: PON-82 added subs checkout page, added checkout skeleton

feat: PON-82 add data layer for new subscription checkout page

feat: PON-82 update subs component with dynamic content rendering

refactor: PON-82 refactored better names for subscription details

feat: PON-82 resolved all PR comments made by Juliana

feat: PON-82 resolved follow up PR comments made by Juliana

julianajlk commented on src/subscription/details/content/SubscriptionContent.jsx:
```diff
@@ -0,0 +1,60 @@
+      <FormattedMessage
+        id="subscription.purchase.details.heading"
+        defaultMessage="In your cart"
+        description="Describe what's in your details"
```
I find this description a little confusing, perhaps something similar to the CartContent component would be clear? 
`Heading of the cart in the subscription details section`

mrazadar replied:
Updated.

julianajlk commented on src/subscription/SubscriptionPage.jsx:
```diff
@@ -69,49 +77,27 @@ export class SubscriptionPage extends React.Component {
+        summarySubtotal={summarySubtotal}
+      />
+      {renderContent()}
+      <ConfirmationModal />
```
Should this component only render on a specific condition? (Also added a comment at the component level)

mrazadar replied:
passed `isVisible` prop from parent that can handle visibility in case of successful payment flow later.

julianajlk commented on src/subscription/data/details/sagas.js:
```diff
@@ -0,0 +1,177 @@
+    yield put(subscriptionDetailsProcessing(true));
+    yield put(clearMessages()); // Don't leave messages floating on the page after clicking submit
+    yield put(submitPayment.request());
+    const paymentMethodCheckout = paymentMethods[method];
```
As commented in another place, we'll only have 1 payment method, so no need for this logic to map out the method to the respective checkout function

mrazadar replied:
updated to use `subscriptionStripeCheckout` method but leave the implementation as it is, so we can easily add more methods in future. 
```
import * as SubscriptionApiService from '../service';
import { subscriptionStripeCheckout } from '../../subscription-methods';

export const paymentMethods = {
  stripe: subscriptionStripeCheckout,
};
```

julianajlk replied:
Sounds good

mrazadar replied:
Sounds good!

julianajlk commented on src/subscription/data/details/selectors.js:
```diff
@@ -0,0 +1,76 @@
+  actualAmount: state[storeName].details.orderTotal,
+});
+
+export const updateQuantityFormSelector = createSelector(
```
Was this selector copied from the regular flow? We won't have a quantity form for subscriptions

mrazadar replied:
removed.

julianajlk commented on src/subscription/data/details/selectors.js:
```diff
@@ -0,0 +1,76 @@
+  subscriptionDetailsSelector,
+  queryParamsSelector,
+  (details, queryParams) => {
+    const isCouponRedeemRedirect = !!queryParams
```
I don't think this is applicable for subscriptions

mrazadar replied:
removed

julianajlk commented on src/subscription/data/service.js:
```diff
@@ -0,0 +1,50 @@
+
+export async function getCaptureKey() {
+  const { data } = await getAuthenticatedHttpClient()
+    .get(`${getConfig().ECOMMERCE_BASE_URL}/bff/payment/v0/capture-context/`)
```
This is not needed, it's for CyberSource

julianajlk commented on src/subscription/data/service.js:
```diff
@@ -0,0 +1,50 @@
+
+export async function getClientSecret() {
+  const { data } = await getAuthenticatedHttpClient()
+    .get(`${getConfig().ECOMMERCE_BASE_URL}/bff/payment/v0/capture-context`)
```
Can you add a TODO comment here to modify the URL when we have the BE subs URL?

julianajlk commented on src/subscription/data/service.js:
```diff
@@ -0,0 +1,50 @@
+  return data;
+}
+
+export async function getDetails(discountJwt) {
```
For this `discountJwt`, I don't believe it's applicable to subscriptions, but will check!

mrazadar replied:
Let's PL this.

julianajlk replied:
Ok this is for the First Purchase Discount that was implemented as a feature, but as Colin mentioned in stand up, we will not have any discounts in subscriptions

mrazadar replied:
removed

julianajlk commented on src/subscription/data/service.js:
```diff
@@ -0,0 +1,50 @@
+export async function getDetails(discountJwt) {
+  const discountJwtArg = typeof discountJwt !== 'undefined' ? `?discount_jwt=${discountJwt}` : '';
+  const { data } = await getAuthenticatedHttpClient()
+    .get(`${getConfig().ECOMMERCE_BASE_URL}/bff/payment/v0/payment/${discountJwtArg}`)
```
Can you add a TODO comment here as well to replace with correct URL?

mrazadar replied:
Done.

julianajlk commented on src/subscription/data/details/selectors.js:
```diff
@@ -0,0 +1,76 @@
+
+import { storeName } from '../constants';
+
+export const localizedCurrencySelector = () => {
```
Could this move to another file where the regular payment flow could also used since it's the same repeated code?

mrazadar replied:
moved under `src/payment/data/utils`

julianajlk commented on src/subscription/data/service.js:
```diff
@@ -0,0 +1,50 @@
+ensureConfig([
+  'ECOMMERCE_BASE_URL',
+  'LMS_BASE_URL',
+], 'payment API service');
```
Can you add a TODO to modify with correct API later?

mrazadar replied:
done.

julianajlk commented on src/subscription/data/service.js:
```diff
@@ -0,0 +1,50 @@
+  'LMS_BASE_URL',
+], 'payment API service');
+
+function handleDetailsApiError(requestError) {
```
Is a different but repeated handle function here for handling errors necessary or could we use the same as regular payment flow?

mrazadar replied:
I think we should keep it, we are not sure what error structure would be from the new Subs BE service. it will be easier to modify it according to our needs in this way.

julianajlk replied:
That's a fair point, we don't know what the BE error structure will be like. In this case, can you either add a TODO or leave the function empty? Because I'd rather not have code in there that we don't know if it works.

mrazadar replied:
added TODO

julianajlk commented on src/subscription/data/details/sagas.js:
```diff
@@ -0,0 +1,177 @@
+import { checkout as checkoutApplePay } from '../../../payment/payment-methods/apple-pay';
+import { checkout as checkoutStripe } from '../../../payment/payment-methods/stripe';
+
+export const paymentMethods = {
```
We will only have Stripe as payment method

mrazadar replied:
Done. 
```js

export const paymentMethods = {
  stripe: subscriptionStripeCheckout,
};
 
```

julianajlk commented on src/subscription/data/details/sagas.js:
```diff
@@ -0,0 +1,177 @@
+import { checkoutWithToken } from '../../../payment/payment-methods/cybersource';
+import { checkout as checkoutPaypal } from '../../../payment/payment-methods/paypal';
+import { checkout as checkoutApplePay } from '../../../payment/payment-methods/apple-pay';
+import { checkout as checkoutStripe } from '../../../payment/payment-methods/stripe';
```
For subscriptions it will be a different checkout function

mrazadar replied:
done

julianajlk reviewed (commented):
Thanks for these changes and refactors! I wonder if we need to repeat some of this code, for example in error handling, and also in the client secret actions/reducer/saga files.

For the loading skeleton of the entire page that you shared the screenshot, nice work! One small detail is that there should be 3 divs at the credit card form and the submit button looks wide.

mrazadar commented on src/payment/cart/CartSkeleton.jsx:
```diff
@@ -1,7 +1,7 @@
 import { FormattedMessage } from '@edx/frontend-platform/i18n';
 
-const SummarySkeleton = () => (
+const CartSkeleton = () => (
```
Because the file name was `CartSkeleton` and it was used as `CatSkeleton` as default export.

mrazadar commented on src/payment/checkout/payment-form/StripePaymentForm.jsx:
```diff
@@ -129,42 +131,45 @@ const StripePaymentForm = ({
-
-    </form>
+        {/* TODO: update onSubmitButtonClick handler for SubscriptionSubmitButton */}
+        {isSubscription ? (
```
yes, its defined in the props. and it has been passed from the parent [here](https://github.com/openedx/frontend-app-payment/blob/c259b18bb0fc70942e90d98b32bcb1e4f5a9b60d/src/subscription/checkout/SubscriptionCheckout.jsx#L95)

mrazadar commented on src/payment/checkout/payment-form/StripePaymentForm.test.jsx:
```diff
@@ -88,6 +89,8 @@ describe('<StripePaymentForm />', () => {
       form: {},
     };
 
+    paymentDataSelector = () => state.payment.basket;
```
done.

mrazadar commented on src/subscription/SubscriptionPage.jsx:
```diff
@@ -1,45 +1,56 @@
+import {
+  FormattedMessage, useIntl,
+} from '@edx/frontend-platform/i18n';
+// import { AppContext } from '@edx/frontend-platform/react';
```
removed the commented code, we were just connecting with context but not consuming it here.

mrazadar commented on src/subscription/SubscriptionPage.jsx:
```diff
@@ -1,45 +1,56 @@
-  renderContent() {
-    const { isEmpty, isRedirect } = this.props;
+  const renderContent = () => {
+    // const { isEmpty, isRedirect } = this.props;
```
removed

mrazadar commented on src/subscription/checkout/SubscriptionCheckout.jsx:
```diff
@@ -0,0 +1,104 @@
+  };
+
+  const handleSubmitStripeButtonClick = () => {
+    sendTrackEvent(
```
this is not final according to Shafqat, so I have added TODO comment with the reference link for now.

mrazadar commented on src/subscription/checkout/SubscriptionCheckout.jsx:
```diff
@@ -0,0 +1,104 @@
+  const intl = useIntl();
+
+  const handleSubmitStripe = (formData) => {
+    dispatch(submitPayment({ method: 'stripe', ...formData }));
```
I have copied the stripe checkout method for now and added couple of TODO's there to update the endpoints later. added `subscription/subscription-methods/stripe` for stripe checkout related to subs

mrazadar commented on src/subscription/checkout/SubscriptionCheckout.jsx:
```diff
@@ -0,0 +1,104 @@
+
+  // TODO: Right now when fetching capture context, CyberSource's captureKey is saved as clientSecretId
+  // so we cannot rely on !options.clientSecret to distinguish btw payment processors
+  const shouldDisplayStripePaymentForm = !loading && enableStripePaymentProcessor && options.clientSecret;
```
removed.

mrazadar commented on src/subscription/checkout/SubscriptionCheckout.jsx:
```diff
@@ -0,0 +1,104 @@
+            paymentDataSelector={subscriptionDetailsSelector}
+          />
+        </Elements>
+      ) : (loading && (<CheckoutSkeleton />))}
```
added a check for `isClientSecretProcessing` 
```js
  ) : ((loading || isClientSecretProcessing) && (<CheckoutSkeleton />))}
```

mrazadar commented on src/subscription/checkout/monthly-billing-notification/MonthlyBillingNotification.jsx:
```diff
@@ -0,0 +1,31 @@
+const messages = defineMessages({
+  'subscription.checkout.billing.notification': {
+    id: 'subscription.checkout.billing.notification',
+    defaultMessage: 'You’ll be charged {price} USD on MM/DD then every 31 days until you cancel your subscription.',
```
should we calculate `MM/DD` by adding 7 days to current date?
what will be the message in case of resubscribe? 

or we can just change the notification to be more generic not mentioning dates. 

e.g. "You be charged $111.0 USD after 7 days and then every 31 days until you cancel your subscription."

julianajlk replied:
This is actually a good option, I will add this in the discussion in Slack

mrazadar commented on src/subscription/checkout/submit-button/SubscriptionSubmitButton.jsx:
```diff
@@ -42,6 +43,13 @@ const SubscriptionSubmitButton = ({
+              resubscribe: (
+                <FormattedMessage
+                  id="subscription.checkout.form.submit.button.text.resubscribe"
+                  defaultMessage="Start my subscription"
```
I think we will have different text for `MonthlyBillingNotification` added a TODO to handle the resubscribe use-case. We can parking lot this today.

julianajlk replied:
Let's not forget to ask this again on Monday!

mrazadar commented on src/subscription/confirmation-modal/ConfirmationModal.jsx:
```diff
@@ -0,0 +1,86 @@
+  // TODO: update the correct URL
+  const ordersAndSubscriptionLink = (
+    <Hyperlink
+      destination={getConfig().ORDER_HISTORY_URL}
```
I think you documented the ORDER_HISTORY_URL under the ticket when I asked this question in one of standup. However we still need redirect URL for `Goto Dashbaord` button, updated the above TODO with this text 
`// TODO: add the redirect URL logic for `Goto Dashboard` button`

julianajlk replied:
Sorry I misunderstood your TODO comment. Noted! 
For the dashboard, is this the learner dashboard or the program dashboard? If program, you could use something like:
`getConfig().LMS_BASE_URL` plus concat `dashboard/programs/`

mrazadar replied:
I have noted this as well and will discuss this in standup.

mrazadar commented on src/subscription/confirmation-modal/ConfirmationModal.jsx:
```diff
@@ -0,0 +1,86 @@
+ * ConfirmationModal
+ */
+export const ConfirmationModal = () => {
+  const [isOpen] = useState(false);
```
my understanding is this will always be `true` and user won't be able to hide it. And we can conditionally render it according to the successful payment flow once that is done.

mrazadar commented on src/subscription/data/details/sagas.js:
```diff
@@ -0,0 +1,177 @@
+    yield put(stopSubmit('payment', fieldErrors));
+  }
+}
+// export function* handleDiscountCheck() {
```
removed.

mrazadar commented on src/subscription/data/details/sagas.js:
```diff
@@ -0,0 +1,177 @@
+  }
+}
+
+// export function* handleAddCoupon({ payload }) {
```
removed

mrazadar commented:
> Thanks for these changes and refactors! I wonder if we need to repeat some of this code, for example in error handling, and also in the client secret actions/reducer/saga files.
> 
> For the loading skeleton of the entire page that you shared the screenshot, nice work! One small detail is that there should be 3 divs at the credit card form and the submit button looks wide.

Updated the skeleton. 
Added replies to the comments. Please have a look and resolve them as per your convenience.
